### ISSUE TITLE
ipv6: startup failed, posix_listen failed for address [fe80::1a:1bff:fe5c:67e6]

### ISSUE THREAD

**Issue by 6Sack** [scylladb/scylla#5177]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 666.development-0.20191013.ef9d5750c82
Cluster size: 1 Node
OS (RHEL/CentOS/Ubuntu/AWS AMI): CentOS Linux release 7.5.1804

I'm trying to test IPV6 support for a single node setup, I have the following parameters in scylla.yaml:

listen_interface: ens3
listen_interface_prefer_ipv6: true
listen_address: fe80::1a:1bff:fe5c:67e6
seeds: "fe80::1a:1bff:fe5c:67e6"

The interface ens3 is setup as a link-local ipv6 address and I have tested that internally it pings and telnets to certain ports on the machine.

When starting scylla the following way, I see the error when requesting the status of the process:

sudo systemctl start scylla-server
sudo systemctl -l status scylla-server

ip-172-31-21-158.us-west-2.compute.internal scylla[11912]:  [shard 0] init - Startup failed: std::system_error (error system:19, posix_listen failed for address [fe80::1a:1bff:fe5c:67e6]:7

When replacing the IPV6 addresses with IPV4 addresses everything works.

**Comment by slivne** [comment]
Thanks @6Sack I think this may be a duplidate of https://github.com/scylladb/scylla/issues/5175 

Scylla does not support the listen_interface as such the listen_interface_prefer_ipv6 is not supported as well (can you please try to comment those out) and try to boot and see if this works for you.

If it doesn't it will be the same as 5175

**Comment by 6Sack** [comment]
Thanks so much for your reply @slivne, I did see that ticket, and did try the same suggestions but instead I just get it to fail sooner during startup:

[centos@ip-172-31-21-158 ~]$ sudo systemctl restart scylla-server
Job for scylla-server.service failed because the control process exited with error code. See "systemctl status scylla-server.service" and "journalctl -xe" for details.

sudo systemctl -l status scylla-server
...
Oct 15 20:19:20 ip-172-31-21-158.us-west-2.compute.internal scylla[3175]:  [shard 0] init - Bad configuration: invalid 'listen_address': fe80::1a:1bff:fe5c:67e6: std::system_error (error C-Ares:4, fe80::1a:1bff:fe5c:67e6: Not found)

I also tried to add %ens3 at the end of the ipv6 address since this is the required way of using link-local addresses, for example when using ssh or ping6 but it produces the same error:

Oct 15 20:19:45 ip-172-31-21-158.us-west-2.compute.internal scylla[3220]:  [shard 0] init - Bad configuration: invalid 'listen_address': fe80::1a:1bff:fe5c:67e6%ens3: std::system_error (error C-Ares:4, fe80::1a:1bff:fe5c:67e6%ens3: Not found)

My understanding is that this was supported in the Master branch, but I'm wondering if it is an issue with using private local-link ipv6 addresses vs global ipv6 addresses, is this a known problem?

Thanks!

**Comment by slivne** [comment]
you do need to set enable-ipv6-dns-lookup to true

**Comment by 6Sack** [comment]
@slivne , that got me further actually, but now this is the error:

enable_ipv6_dns_lookup: true

Oct 15 20:31:46 ip-172-31-21-158.us-west-2.compute.internal scylla[3428]:  [shard 0] init - Startup failed: std::system_error (error system:19, posix_listen failed for address [fe80::1a:1bff:fe5c:67e6]:7000: No such device)

I tried with %ens3 but that doesn't make it that far.

**Comment by slivne** [comment]
Well at least for me the basic test is working

```
netstat -an | more

Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN     
tcp        0      0 127.0.0.1:25            0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:9180            0.0.0.0:*               LISTEN     
tcp        0      0 0.0.0.0:111             0.0.0.0:*               LISTEN     
tcp        0      0 127.0.0.1:10000         0.0.0.0:*               LISTEN     
tcp        0     36 10.0.0.251:22           5.29.134.155:54716      ESTABLISHED
tcp        1      0 10.0.0.251:62618        169.254.169.254:80      CLOSE_WAIT 
tcp6       0      0 2600:1f18:26b1:3a0:9042 :::*                    LISTEN     
tcp6       0      0 :::22                   :::*                    LISTEN     
tcp6       0      0 2600:1f18:26b1:3a0:7000 :::*                    LISTEN     
tcp6       0      0 :::39545                :::*                    LISTEN     
tcp6       0      0 ::1:25                  :::*                    LISTEN     
tcp6       0      0 127.0.0.1:7199          :::*                    LISTEN     
tcp6       0      0 2600:1f18:26b1:3a0:9160 :::*                    LISTEN     
tcp6       0      0 :::9100                 :::*                    LISTEN     
tcp6       0      0 :::111                  :::*                    LISTEN     
tcp6       0      0 2600:1f18:26b1:3a:42732 2600:1f18:26b1:3a0:9042 TIME_WAIT  
tcp6       0      0 2600:1f18:26b1:3a:42730 2600:1f18:26b1:3a0:9042 TIME_WAIT  
udp        0      0 0.0.0.0:68              0.0.0.0:*                          
udp        0      0 0.0.0.0:111             0.0.0.0:*                          
udp        0      0 127.0.0.1:323           0.0.0.0:*                          
udp6       0      0 :::111                  :::*                               
udp6       0      0 ::1:323                 :::*                               
udp6       0      0 fe80::844:5bff:fe32:546 :::*                               
Active UNIX domain sockets (servers and established)
```

```
cat  /etc/scylla/scylla.yaml  | grep 2600:1f18:26b1:3a00:3117:b641:b849:f78
          - seeds: "2600:1f18:26b1:3a00:3117:b641:b849:f78"
listen_address: 2600:1f18:26b1:3a00:3117:b641:b849:f78
rpc_address: 2600:1f18:26b1:3a00:3117:b641:b849:f78
broadcast_rpc_address: 2600:1f18:26b1:3a00:3117:b641:b849:f78
```

I started from the scylla-ami and then changed the config, cleaned the /var/lib/scylla/.../ and booted and it works

so I changed seeds / listen_address / rpc_address / broadcast_rpc_address

the rest api is still on 127.0.0.1 and we need to test this as well ...

I was even able to test cqlsh
```
cqlsh 2600:1f18:26b1:3a00:3117:b641:b849:f78
Connected to Test Cluster - No AMI Parameters at 2600:1f18:26b1:3a00:3117:b641:b849:f78:9042.
[cqlsh 5.0.1 | Cassandra 3.0.8 | CQL spec 3.3.1 | Native protocol v4]
Use HELP for help.
cqlsh> 
```

and also cassandra-stress
```
 cassandra-stress write n=10000 -node 2600:1f18:26b1:3a00:3117:b641:b849:f78

Connected to cluster: Test Cluster - No AMI Parameters, max pending requests per connection 128, max connections per host 8
Datatacenter: us-east; Host: /2600:1f18:26b1:3a00:3117:b641:b849:f78; Rack: 1c
Created keyspaces. Sleeping 1s for propagation.
Sleeping 2s...
Warming up WRITE with 2500 iterations...
Failed to connect over JMX; not collecting these stats
Running WRITE with 200 threads for 10000 iteration
Failed to connect over JMX; not collecting these stats
type       total ops,    op/s,    pk/s,   row/s,    mean,     med,     .95,     .99,    .999,     max,   time,   stderr, errors,  gc: #,  max ms,  sum ms,  sdv ms,      mb
total,          3029,    3029,    3029,    3029,    48.4,    44.4,   102.4,   133.4,   160.0,   162.8,    1.0,  0.00000,      0,      0,       0,       0,       0,       0
```

So ipv6 does work at least it worked for me

@6Sack I need the exact setting you did the complete scylla.yaml

**Comment by slivne** [comment]
for reference I also set ```enable_ipv6_dns_lookup: true```

**Comment by 6Sack** [comment]
Let me try with global ipv6 addresses, I have a feeling this is a problem with link-local addresses only. Will get back to you.

**Comment by elcallio** [comment]
You are using a link-local address. On linux, binding to a link-local address requires specifying the "scope" (i.e. the interface). The problem is that we don't use IPV6 address scope properly. In fact, we don't even properly set it to zero as a default. 

For your case, since the address is in fact tied to scope zero, having a default scope set in binding would fix your issue. However, iff someone tries to bind to a link-local address on a different scope, we don't really have a good way to communicate this down the stack. Scope is _not_ part of addresses, but more akin to ports.

**Comment by 6Sack** [comment]
Correct, that is what I speculated since there was nowhere to specify the scope/interface for link-local addresses and the '%' notation was not being allowed. 

After trying global ipv6 addresses I got over this issue, but now I have the same problem reported in #5175, so I will switch context to that:

Oct 16 06:07:55 ip-172-31-17-93.us-west-2.compute.internal scylla[4838]:  [shard 0] rpc - client [2600:1f14::]:7000: fail to connect: bind: Address family not supported by protocol

The only open question for this ticket is if we believe we should support link-local ipv6 addresses. I believe we perhaps should, but will leave this to the community.

**Comment by 6Sack** [comment]
@slivne btw, just to close the loop, let me give you some info on my configuration and what I found out:

```
cat  /etc/scylla/scylla.yaml | grep 2600:1f14:e86:2703:3f22:ec7d:5fbf:44e7
          - seeds: "2600:1f14:e86:2703:3f22:ec7d:5fbf:44e7,2600:1f14:e86:2703:a1a9:a99b:3088:c5b6,2600:1f14:e86:2703:bc64:9052:2ae:5dd5"
listen_address: 2600:1f14:e86:2703:3f22:ec7d:5fbf:44e7
rpc_address: 2600:1f14:e86:2703:3f22:ec7d:5fbf:44e7
broadcast_rpc_address: 2600:1f14:e86:2703:3f22:ec7d:5fbf:44e7
```

and the other following settings commented out:
```
#listen_interface: ens3
#listen_interface_prefer_ipv6: true
```

and this setting enabled:
```
enable_ipv6_dns_lookup: true
```

When I tested it on a single node, it works, when trying to create a cluster, it does not work and the logs have the following:
```
Oct 16 19:34:33 ip-172-31-21-158.us-west-2.compute.internal scylla[23641]:  [shard 0] rpc - client [2600:1f14:e86:2703:bc64:9052:2ae:5dd5]:7000: fail to connect: bind: Address family not supported by protocol
```

nodetool status:
```
UN  2600:1f14:e86:2703:3f22:ec7d:5fbf:44e7  766.4 KB   256          ?       e919feab-f67e-4cdb-bdfc-238178244749  rack1
```

It eventually fails to start Scylla. Similar behavior as #5175.

**Comment by aleksbykov** [comment]
On this version 3.1.0-0.20191017.9c3cdded9
always get this error:
```
окт 18 09:53:06 ip-10-0-58-8.eu-west-1.compute.internal scylla[1842]: Scylla version 3.1.0-0.20191017.9c3cdded9 starting ...
окт 18 09:53:06 ip-10-0-58-8.eu-west-1.compute.internal scylla[1842]:  [shard 0] init - Scylla version 3.1.0-0.20191017.9c3cdded9 starting.
Oct 18 09:53:06 ip-10-0-58-8.eu-west-1.compute.internal scylla[1842]:  [shard 0] init - Bad configuration: invalid 'listen_address': 2a05:d018:223:f00:f271:3ef3:df2e:fe26: std::system_error (err... Not found)
Oct 18 09:53:06 ip-10-0-58-8.eu-west-1.compute.internal systemd[1]: scylla-server.service: main process exited, code=exited, status=1/FAILURE
Oct 18 09:53:06 ip-10-0-58-8.eu-west-1.compute.internal systemd[1]: Failed to start Scylla Server.
```

**Comment by slivne** [comment]
patch merged to scylla 771e028c1aafb86a870fbffd845cbb8fa2f93564
